paginator/paginator.go

```diff
@@ -259,6 +267,46 @@ func (p *Paginator) buildCursorSQLQuery() string {
 	return strings.Join(queries, " OR ")
 }
 
+// We can only optimize paging query if sorting orders are consistent across
+// all columns used in cursor. This is a prerequisite for tuple comparison that
+// optimized queries use.
+func (p *Paginator) canOptimizePagingQuery() bool {
+	order := p.rules[0].Order
+
+	for _, rule := range p.rules {
+		if order != rule.Order {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *Paginator) getCmpOperator(order Order) string {
+	if (p.isForward() && order == ASC) || (p.isBackward() && order == DESC) {
+		return ">"
+	}
+
+	return "<"
+}
+
+func (p *Paginator) buildOptimizedCursorSQLQuery() string {
+	names := make([]string, len(p.rules))
+	values := make([]string, len(p.rules))
+
+	for i, rule := range p.rules {
+		names[i] = rule.SQLRepr
```

Comment:
SQLRepr already has null replacement & type casting handled, right?

Reply:
Yes, all that happens in the `setup` call. This part of functionality is shared with the original method of generating query SQL.